File: src/main/python/AoC2016_07.py
#! /usr/bin/env python3
#
# Advent of Code 2015 Day 7
#

import re
import sys

from aoc.common import InputData
from aoc.common import SolutionBase
from aoc.common import aoc_samples

ABBA = re.compile(r"([a-z])(?!\1)([a-z])\2\1")
HYPERNET = re.compile(r"\[([a-z]*)\]")
ABA = re.compile(r"([a-z])(?!\1)[a-z]\1")

TEST1 = """\
abba[mnop]qrst
abcd[bddb]xyyx
aaaa[qwer]tyui
ioxxoj[asdfgh]zxcvbn
abcox[ooooo]xocba
"""
TEST2 = """\
aba[bab]xyz
xyx[xyx]xyx
aaa[kek]eke
zazbz[bzb]cdb
"""

Input = InputData
Output1 = int
Output2 = int


class Solution(SolutionBase[Input, Output1, Output2]):
    def parse_input(self, input_data: InputData) -> Input:
        return input_data

    def part_1(self, inputs: InputData) -> int:
        def is_tls(ip: str) -> bool:
            for b in HYPERNET.findall(ip):
                if ABBA.search(b) is not None:
                    return False
                ip = ip.replace("[" + b + "]", "/")
            return ABBA.search(ip) is not None

        return sum(is_tls(line) for line in inputs)

    def part_2(self, inputs: InputData) -> int:
        def is_sls(ip: str) -> bool:
            bs = HYPERNET.findall(ip)
            for b in bs:
                ip = ip.replace("[" + b + "]", "/")
            return any(
                any(bab in b for b in bs)
                for bab in (
                    m.group()[1] + m.group()[0] + m.group()[1]
                    for i in range(len(ip) - 3)
                    for m in ABA.finditer(ip[i:])
                )
            )

        return sum(is_sls(line) for line in inputs)

    @aoc_samples(
        (
            ("part_1", TEST1, 2),
            ("part_2", TEST2, 3),
        )
    )
    def samples(self) -> None:
        pass


solution = Solution(2016, 7)


def main() -> None:
    solution.run(sys.argv)


if __name__ == "__main__":
    main()

File: src/main/java/AoC2016_07.java
import static com.github.pareronia.aoc.Utils.toAString;

import java.util.List;
import java.util.regex.Pattern;
import java.util.stream.Stream;

import com.github.pareronia.aoc.solution.Sample;
import com.github.pareronia.aoc.solution.Samples;
import com.github.pareronia.aoc.solution.SolutionBase;

public class AoC2016_07 extends SolutionBase<List<String>, Integer, Integer> {

    private static final Pattern ABBA = Pattern.compile("([a-z])(?!\\1)([a-z])\\2\\1");
    private static final Pattern HYPERNET = Pattern.compile("\\[([a-z]*)\\]");
    private static final Pattern ABA = Pattern.compile("([a-z])(?!\\1)[a-z]\\1");

	private AoC2016_07(final boolean debug) {
		super(debug);
	}
	
	public static final AoC2016_07 create() {
		return new AoC2016_07(false);
	}

	public static final AoC2016_07 createDebug() {
		return new AoC2016_07(true);
	}
	
	@Override
    protected List<String> parseInput(final List<String> inputs) {
        return inputs;
    }

    private boolean isTLS(final String ip) {
	    final List<String> mhs = HYPERNET.matcher(ip).results()
	            .map(m -> m.group(1))
	            .toList();
	    String temp = ip;
	    for (final String mh : mhs) {
	        if (ABBA.matcher(mh).find()) {
	            return false;
	        }
	        temp = temp.replace("[" + mh + "]", "/");
	    }
	    return ABBA.matcher(temp).find();
	}
	
	private String aba2bab(final String string) {
	    assert string.length() == 3;
	    return Stream.of(string.charAt(1), string.charAt(0), string.charAt(1))
	            .collect(toAString());
	}
	
	private boolean isSLS(final String ip) {
	    final List<String> mhs = HYPERNET.matcher(ip).results()
	            .map(m -> m.group(1))
	            .toList();
	    String temp = ip;
	    for (final String mh : mhs) {
	        temp = temp.replace("[" + mh + "]", "/");
	    }
	    final String temp2 = temp;
	    return Stream.iterate(0, i -> i + 1)
	            .limit(temp2.length() - 3)
	            .flatMap(i -> ABA.matcher(temp2.substring(i)).results())
	            .map(m -> m.group(0))
	            .map(this::aba2bab)
	            .anyMatch(aba -> mhs.stream().anyMatch(mh -> mh.contains(aba)));
	}
	
	@Override
	public Integer solvePart1(final List<String> inputs) {
		return (int) inputs.stream().filter(this::isTLS).count();
	}

	@Override
	public Integer solvePart2(final List<String> inputs) {
		return (int) inputs.stream().filter(this::isSLS).count();
	}

	@Samples({
	    @Sample(method = "part1", input = TEST1, expected = "2"),
	    @Sample(method = "part2", input = TEST2, expected = "3"),
	})
	public static void main(final String[] args) throws Exception {
	    AoC2016_07.create().run();
	}

	private static final String TEST1 = """
	        abba[mnop]qrst\r
	        abcd[bddb]xyyx\r
	        aaaa[qwer]tyui\r
	        ioxxoj[asdfgh]zxcvbn\r
	        abcox[ooooo]xocba
	        """;
	private static final String TEST2 = """
	        aba[bab]xyz\r
	        xyx[xyx]xyx\r
	        aaa[kek]eke\r
	        zazbz[bzb]cdb
	        """;
}
